3175 remove algolia dependency

entantoencuanto opened the pull request:
Closes #3175

## :v: What does this PR do?

Moves search engine from algolia to pg_search:
* Configures pg_search with multisearch option and adds some attributes and associations to the `PgSearch::Document` model:
  * `belongs_to` association with `Site`
  * Translatable attributes: `title` and `description`
  * `resource_path`
  * `meta`: Used to store some extra attributes like kind of budget_line or collection title and id of GobiertoCms::Page instances
  * `external_id` used in `GobiertoBudgets::BudgetLine` document types because the indexed elements are not stored in database
  * `searchable_udpated_at`: It copies the `updated_at` attribute from searchable
* Adds multisearch to models indexed with algolia
* Extends search support to `GobiertoPlans::Node` taking into account versions and including in search document custom fields of type text values
* Adds some tests for plans projects checking that the search documents contains the versioned values of the native fields and the custom fields
* Replaces the concern used to index budget lines in algolia to do the same job in pg_search and changes callbacks using algolia to use pg_search instead
* Adds an API endpoint to make searches using pg_search.
* Replaces global front search box and pages search in admin/cms/sections to use the API instead of algolia
* Removes the algolia js package and gem
* Fixes styles of front search box
* Removes derprecated tasks and fixes some tests

## :mag: How should this be manually tested?


## :eyes: Screenshots

### Before this PR

### After this PR

## :shipit: Does this PR changes any configuration file?

- There is a PR in populate-ansible to remove environment variables related to algolia

(Changes in these files might need to update the role in Ansible)

## :book: Does this PR require updating the documentation?

No

entantoencuanto committed:
Add pg_search gem

entantoencuanto committed:
Update Gemfile.lock

entantoencuanto committed:
Add pg_search_documents table for pg_search multisearch

entantoencuanto committed:
Update db/schema.rb

entantoencuanto committed:
Include pg_search support in searchable concern

entantoencuanto committed:
Add multisearchable to GobiertoCitizensCharter::Commitment model

entantoencuanto committed:
Add migration to enable unaccent and trigram extensions

entantoencuanto committed:
Configure pg_search and define associations and translated attributes

entantoencuanto committed:
Add Site multisearch instance method

entantoencuanto committed:
Add task to reindex all models with pg_search configured

entantoencuanto committed:
Add missing models

entantoencuanto committed:
Allow searchable concern to deal with not translatable attributes for documents title or description

entantoencuanto committed:
Add multisearchable to remaining models using algolia

entantoencuanto committed:
Define search/query API endpoint

entantoencuanto committed:
Rename search_client method and include searchable_types with searchable model names

entantoencuanto committed:
Replace algolia search with autocomplete calling search API

entantoencuanto committed:
Ignore searchable_type param in module-search if searcClient is not present

entantoencuanto committed:
Remove unused powered_by algolia translation

entantoencuanto committed:
Add searchable? method to avoid indexing of archived items

In the case of events the reindexing callback is triggered after archive
an event.

entantoencuanto committed:
Don't return a public path for archived events

ferblape commented on app/javascript/lib/shared/modules/module-search.js:
```diff
@@ -60,73 +63,42 @@ $(document).on('turbolinks:load', function() {
-      $('<div class="result"><small>'+I18n.t("layouts.search.powered_by")+'</small></div>').appendTo($resultsContainer);
-    }
+  function formattedResult(d){
+    return '<h2>' + d['title'] + '</h2>' +
```
You can use interpolations, i.e `<h2>${d['title']}</h2>` + ....

entantoencuanto replied:
Done!

entantoencuanto committed:
Fix param name

entantoencuanto committed:
Add external_id and searchable_updated_at columns to pg_search_documents table

entantoencuanto committed:
Update db/schema.rb

entantoencuanto committed:
Improve regular expression

entantoencuanto committed:
Add helper to index budget lines in pg_search

entantoencuanto committed:
Don't index description for models without descriptions in algolia

Affected models:
* GobiertoCms::Page
* GobiertoCalendars::Event
* GobiertoCitizensCharters::Commitment

entantoencuanto committed:
Index searchable updated_at in search documents instead of the document itself

entantoencuanto committed:
Fix method to generate content of budget lines search documents

entantoencuanto committed:
Add some small front fixes

* Ignore updated at if search API send it with null value
* Take into account the Income/Expense type of budget lines looking in
  meta

entantoencuanto committed:
Add task to regenerate search documents of budget lines by site and year

entantoencuanto committed:
Change custom fields searchable_value

entantoencuanto committed:
Add searchable concern to GobiertoPlans::Node model

The multisearchable call in the model makes use of the serializer
defined for the model which takes into account custom fields and
versions

entantoencuanto committed:
Change update of pg_search document to consider versions

entantoencuanto committed:
Update pg_search_document from custom_field_records_form

This step has to be added because the custom fields are updated after
the resource is saved and reindex callback is executed

entantoencuanto committed:
Add integration test for plan projects and searches reindexing

Crashillo committed:
autocomplete styles

Crashillo commented:
You may take a look @entantoencuanto

entantoencuanto committed:
Prevent errors reindexing nodes with blank published version

On csv import with automatic publication and some tests the
nodes were published before setting the published version and
this was producing an error trying to reindex the node

entantoencuanto committed:
Include GobiertoPlans::Node in searchable models

entantoencuanto committed:
Fix tests

ferblape commented on app/models/concerns/gobierto_budgets/budget_line_algolia_helpers.rb:
```diff
@@ -45,10 +45,10 @@ def algolia_as_json
       end
 
-      def translated_attributes
+      def algolia_translated_attributes
```
Why not remove algolia from the file name and from the method?

entantoencuanto replied:
This entire helper file is going to be removed after cleaning the algolia dependencies, but this method name conflicts with another one added by the new code

entantoencuanto replied:
Al the algolia references has been deleted in this branch

ferblape commented on app/views/layouts/_gobierto_footer.html.erb:
```diff
@@ -12,8 +12,9 @@
   <% if algoliasearch_configured? %>
     window.searchClient = {
       client: algoliasearch('<%= Rails.application.secrets.algolia_application_id %>', '<%= Rails.application.secrets.algolia_search_api_key %>'),
```
can you remove the client option, right?

entantoencuanto committed:
Avoid exceptions on versionable serializations when published_version is blank

entantoencuanto committed:
Replace strings to use interpolations

entantoencuanto committed:
Replace algolia with pg_search to reindex budget lines and update test

entantoencuanto committed:
Replace algolia searches in admin/cms/sections

entantoencuanto committed:
Add meta to gobierto_cms pg_search documents

entantoencuanto committed:
Remove algolia from budget lines

entantoencuanto committed:
Add extra info to attachment search documents meta

entantoencuanto committed:
Remove algoliasearch_gobierto from models and concern

entantoencuanto committed:
Replace trigger_reindex_job with pg_search method

entantoencuanto committed:
Remove algolia from javascripts

entantoencuanto committed:
Update yarn.lock

entantoencuanto committed:
Remove algolia gem, configuration and initializers

entantoencuanto committed:
Update Gemfile.lock

entantoencuanto committed:
Remove algoliasearch_configured? method

entantoencuanto committed:
Remove deprecated algolia_search_disabled? site configuration setting

entantoencuanto committed:
Allow to reindex on task only if no indexes present or forced by option

ferblape commented on app/controllers/gobierto_common/api/v1/search_controller.rb:
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
```
Could you add a test of this endpoint?

entantoencuanto replied:
Done!

entantoencuanto committed:
Merge branch 'master' into 3175-remove_algolia_dependency

* master: (140 commits)
  Bump elliptic from 6.5.1 to 6.5.3
  Update appsignal to version 2.10.12
  Update codecov to version 0.2.3
  scroll pagination
  show pagination as computed property in the component
  resolved changes @crashillo
  Update google-api-client to version 0.42.2
  Add integration test
  Control if charlock returns nil
  Update rollbar to version 2.27.0
  Accept column names prefixed with Node for external_id
  update pagination and number of pages when use the filters
  i18 locales | styles
  change styles
  add new column sumadatos
  styles component
  add another check, use every instead of some
  pagination component
  Detect CSV encoding before UTF-8 conversion
  Update bootsnap to version 1.4.7
  ...

entantoencuanto committed:
Add tests for search API controller

ferblape reviewed (changes requested):
- in Gobierto data module the search input doesn't work
- if you open it in mobile it doesn't work either. You don't need to use a mobile, you can make the browser small to enable the mobile layout

entantoencuanto committed:
Allow budgets reindex task to autodetect last year with data

entantoencuanto committed:
Disable turbolinks in search box results

entantoencuanto committed:
Fix styles

entantoencuanto committed:
Use different id for mobile input search and define its own autocomplete

entantoencuanto committed:
Don't execute searches code on turbolinks:load event

entantoencuanto commented:
> * in Gobierto data module the search input doesn't work

Attached to another event, I think it's working now

> * if you open it in mobile it doesn't work either. You don't need to use a mobile, you can make the browser small to enable the mobile layout

There was a duplication in the input id, I've set two different inputs with their own autocomplete
